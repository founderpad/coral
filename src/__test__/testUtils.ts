// import { ThemeProvider } from "my-ui-lib"
// import { TranslationProvider } from "my-i18n-lib"
// import defaultStrings from "i18n/en-x-default"
import { render } from '@testing-library/react';
import { JSXElementConstructor, ReactElement } from 'react';

const Providers = ({ children }: { children: any }) => {
	return children;
	// return (
	//   <ThemeProvider theme="light">
	//     <TranslationProvider messages={defaultStrings}>
	//       {children}
	//     </TranslationProvider>
	//   </ThemeProvider>
	// )
};

const customRender = (
	ui: ReactElement<any, string | JSXElementConstructor<any>>,
	options = {}
): any => render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
