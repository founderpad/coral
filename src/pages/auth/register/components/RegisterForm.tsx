import { FormInput } from '@/components/form/inputs/FormField';
import { useRegister } from '@/hooks/auth';
import React, { useContext } from 'react';
import LegalFooter from './LegalFooter';
import { TRegisterFormFields } from '../../../../types/auth';
import { BaseForm } from '@/components/form';
import { SubmitButton } from '@/components/buttons';
import { AlertFeedback } from '@/components/alert';
import SocialLogins from '@/components/shared/SocialLogins';
import NotificationContext from '@/context/NotificationContext';
import { schema } from '../validationSchema';

const defaultValues: Record<string, string> & TRegisterFormFields = {
	firstName: '',
	lastName: '',
	email: '',
	password: ''
};

const RegisterForm = () => {
	const { onRegister } = useRegister();
	const { notification } = useContext(NotificationContext);

	return (
		<React.Fragment>
			<BaseForm<TRegisterFormFields, typeof schema>
				name="register-form"
				onSubmit={onRegister}
				defaultValues={defaultValues}
				schema={schema}
				stackProps={{
					alignItems: 'center',
					spacing: 3
				}}
			>
				{({
					register,
					control,
					formState: { errors, isSubmitting }
				}) => (
					<React.Fragment>
						<FormInput<TRegisterFormFields>
							id="firstName"
							name="firstName"
							placeholder="First name"
							register={register}
							control={control}
							errors={errors}
							fieldProps={{
								placeholder: 'First name'
							}}
							hideLimit={true}
							hideClear
						/>

						<FormInput<TRegisterFormFields>
							id="lastName"
							name="lastName"
							placeholder="Last name"
							register={register}
							control={control}
							fieldProps={{
								placeholder: 'Last name'
							}}
							hideLimit={true}
							errors={errors}
							hideClear
						/>
						<FormInput<TRegisterFormFields>
							id="email"
							name="email"
							placeholder="Email"
							register={register}
							control={control}
							fieldProps={{
								placeholder: 'Email'
							}}
							hideLimit={true}
							errors={errors}
							hideClear
						/>

						<FormInput<TRegisterFormFields>
							id="password"
							name="password"
							placeholder="Password"
							register={register}
							control={control}
							fieldProps={{
								placeholder: 'Password',
								type: 'password'
							}}
							hideLimit={true}
							errors={errors}
							hideClear
						/>

						{notification && (
							<AlertFeedback
								status={notification.status}
								message={notification.message}
							/>
						)}

						<SubmitButton
							id="submit-register-account"
							name="submit-register-account"
							label="Create account"
							isLoading={isSubmitting && !notification}
							disabled={isSubmitting && !notification}
							size="md"
							fontSize="small"
							w={{ base: 'full', sm: '200px' }}
						/>
					</React.Fragment>
				)}
			</BaseForm>
			<SocialLogins />

			{/* <Alert fontSize="xs" status="warning" p={2} textAlign="center">
				Your email address must be the same as your PayPal account in
				order to withdraw funds.
			</Alert> */}
			<LegalFooter />
		</React.Fragment>
	);
};

export default RegisterForm;
