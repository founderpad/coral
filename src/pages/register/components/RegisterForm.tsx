import { SubmitButton } from '@components/buttons';
import { Form } from '@components/form';
import {
	EmailField,
	InputField,
	PasswordField
} from '@components/input/InputField';
// import SocialLogins from '@components/shared/SocialLogins';
import { useRegister } from '@hooks/auth';
import React from 'react';
import { useForm } from 'react-hook-form';
import { IRegisterFormData } from 'src/types/auth';
import LegalFooter from './LegalFooter';

const RegisterForm = () => {
	const {
		handleSubmit,
		control,
		formState: { errors, isSubmitting, isValid }
	} = useForm<IRegisterFormData>({ mode: 'all' });
	const onRegister = useRegister();

	return (
		<React.Fragment>
			<Form
				id={'registerForm'}
				name={'registerForm'}
				onSubmit={handleSubmit(onRegister)}
				stackProps={{
					alignItems: 'center'
				}}
			>
				<InputField
					name="firstName"
					error={errors.firstName}
					errorText="You must input a first name"
					placeholder="First name"
					fontSize={'sm'}
					control={control}
					rules={{ maxLength: 20, minLength: 2 }}
					size={'md'}
					isRequired
				/>
				<InputField
					name="lastName"
					placeholder="Last name"
					control={control}
					rules={{ minLength: 0, maxLength: 20 }}
					size={'md'}
					fontSize={'sm'}
				/>
				<EmailField
					id="email"
					name="email"
					error={errors['email']}
					control={control}
					size={'md'}
					fontSize={'sm'}
					isRequired
				/>
				<PasswordField
					id="password"
					name="password"
					error={errors['password']}
					rules={{ maxLength: 20, minLength: 6 }}
					control={control}
					size={'md'}
					fontSize={'sm'}
					isRequired
				/>
				{/* <AlertFeedback /> */}
				<SubmitButton
					id={'submit-register-account'}
					name={'submit-register-account'}
					label={'Create account'}
					isLoading={isSubmitting}
					disabled={!isValid || isSubmitting}
					// disabled={isSubmitting}
					size={'md'}
					fontSize={'sm'}
					w={{ base: 'full', sm: '175px' }}
				/>
			</Form>
			{/* <SocialLogins /> */}
			<LegalFooter />
		</React.Fragment>
	);
};

export default RegisterForm;
