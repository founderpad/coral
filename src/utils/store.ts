import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import { persistReducer } from 'redux-persist';
import { PersistConfig } from 'redux-persist/es/types';
import storage from 'redux-persist/lib/storage';
import rootReducer from './reducer';

/**
 * Create the Redux store
 *
 * @author jlee
 */

// const isProd = process.env.NODE_ENV !== 'development';

/**
 * Persist storage configuration set up
 * @key[root]
 * @storage[storage]
 */
const persistConfig: PersistConfig<any> = {
	key: 'root',
	storage
};

const middleware = [...getDefaultMiddleware({ serializableCheck: false })];

const persistedReducer = persistReducer(persistConfig, rootReducer);

const store = configureStore({
	reducer: persistedReducer,
	middleware
	// devTools: isProd ?? false
});

// export { StoreProvider }

export default store;
