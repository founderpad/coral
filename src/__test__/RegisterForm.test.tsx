import RegisterForm from '@pages/register/components/RegisterForm';
import store from '@utils/store';
import { Provider } from 'react-redux';
import { act, fireEvent, render, userEvent, waitFor } from './testUtils';

const setup = () => {
	const registerSetup = render(
		<Provider store={store}>
			<RegisterForm />
		</Provider>
	);

	const registerForm = registerSetup.getByRole('form', {
		name: /register-form/i
	});
	const firstNameField = registerSetup.getByRole('textbox', {
		name: /firstName/i
	});
	const lastNameField = registerSetup.getByRole('textbox', {
		name: /lastName/i
	});
	const emailField = registerSetup.getByRole('textbox', { name: /email/i });
	const passwordField = registerSetup.getByPlaceholderText(/Password/i);

	const submitButton = registerSetup.getByRole('button', { name: /submit/i });
	// const socialSignup = registerSetup.getByTestId('socialLogin');

	return {
		registerSetup,
		registerForm,
		firstNameField,
		lastNameField,
		emailField,
		passwordField,
		submitButton,
		// socialSignup,
		...registerSetup
	};
};

const mockRegister = jest.fn();
// const mockSocialRegister = jest.fn();

jest.mock('@hooks/auth', () => ({
	useRegister: (): any => mockRegister
	// useSocialLogin: (): any => mockSocialRegister
}));

describe('Register form', () => {
	it('matches snapshot', async () => {
		const { asFragment } = setup();
		await waitFor(() => asFragment());
		expect(asFragment()).toMatchSnapshot();
	});

	it('should render RegisterForm', async () => {
		const { registerForm } = setup();
		expect(registerForm).toBeInTheDocument();
	});

	it('should render RegisterForm fields and submit should be enabled by default', async () => {
		const {
			emailField,
			passwordField,
			firstNameField,
			lastNameField,
			submitButton
		} = setup();
		expect(firstNameField).toBeInTheDocument();
		expect(lastNameField).toBeInTheDocument();
		expect(emailField).toBeInTheDocument();
		expect(passwordField).toBeInTheDocument();
		expect(submitButton).toBeEnabled();
	});

	it('should make successful register request on submit', async () => {
		const {
			firstNameField,
			lastNameField,
			emailField,
			passwordField,
			submitButton
		} = setup();

		expect(firstNameField).toBeInTheDocument();
		expect(lastNameField).toBeInTheDocument();
		expect(emailField).toBeInTheDocument();
		expect(passwordField).toBeInTheDocument();

		userEvent.type(firstNameField, 'Jamie');
		await waitFor(() => expect(firstNameField).toHaveValue('Jamie'));

		userEvent.type(lastNameField, 'Lee');
		await waitFor(() => expect(lastNameField).toHaveValue('Lee'));

		userEvent.type(emailField, 'jamie@gmail.com');
		await waitFor(() => expect(emailField).toHaveValue('jamie@gmail.com'));

		userEvent.type(passwordField, 'admin123456');
		await waitFor(() => expect(passwordField).toHaveValue('admin123456'));

		await act(async () => {
			fireEvent.click(submitButton);
		});

		expect(mockRegister).toHaveBeenCalledTimes(1);
	});

	it('should error if firstName is too short (< 2 characters) on blur', async () => {
		const { registerSetup, firstNameField } = setup();

		expect(firstNameField).toBeInTheDocument();

		fireEvent.focus(firstNameField);
		userEvent.type(firstNameField, 'j');
		await waitFor(() => expect(firstNameField).toHaveValue('j'));

		fireEvent.blur(firstNameField);

		registerSetup.getByText(
			'You first name must be a minimum of 2 characters'
		);
	});

	it('should error if firstName is too long (> 30 characters) on blur', async () => {
		const { registerSetup, firstNameField } = setup();

		expect(firstNameField).toBeInTheDocument();

		fireEvent.focus(firstNameField);
		userEvent.type(firstNameField, 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa');
		await waitFor(() =>
			expect(firstNameField).toHaveValue(
				'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'
			)
		);

		fireEvent.blur(firstNameField);

		registerSetup.getByText(
			'You first name must be a maximum of 30 characters'
		);
	});

	it('should error if lastName is too long (> 30 characters) on blur', async () => {
		const { registerSetup, lastNameField } = setup();

		expect(lastNameField).toBeInTheDocument();

		fireEvent.focus(lastNameField);
		userEvent.type(lastNameField, 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa');
		await waitFor(() =>
			expect(lastNameField).toHaveValue('aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa')
		);

		fireEvent.blur(lastNameField);

		registerSetup.getByText(
			'Your last name must be a maximum of 30 characters'
		);
	});

	it('should error if invalid register email format on blur', async () => {
		const { registerSetup, emailField } = setup();

		expect(emailField).toBeInTheDocument();

		fireEvent.focus(emailField);
		userEvent.type(emailField, 'testemail');
		await waitFor(() => expect(emailField).toHaveValue('testemail'));

		fireEvent.blur(emailField);

		registerSetup.getByText('You must enter a valid email address');
	});

	it('should error if register email is empty on blur', async () => {
		const { registerSetup, emailField, submitButton } = setup();

		expect(emailField).toBeInTheDocument();

		await act(async () => {
			fireEvent.click(submitButton);
		});

		registerSetup.getByText('You must enter a valid email address');
	});

	it('should error if password too short (< 6 characters) on blur', async () => {
		const { registerSetup, emailField, passwordField } = setup();

		expect(emailField).toBeInTheDocument();
		expect(passwordField).toBeInTheDocument();

		fireEvent.focus(passwordField);
		userEvent.type(passwordField, 'pwd');
		await waitFor(() => expect(passwordField).toHaveValue('pwd'));

		fireEvent.blur(passwordField);

		registerSetup.getByText(
			'Your password must be a minimum of 6 characters'
		);
	});

	it('should error if password too long (> 20 characters) on blur', async () => {
		const { registerSetup, emailField, passwordField } = setup();

		expect(emailField).toBeInTheDocument();
		expect(passwordField).toBeInTheDocument();

		fireEvent.focus(passwordField);
		userEvent.type(passwordField, 'abcdefghjklmnopqrstuvwxyz');
		await waitFor(() =>
			expect(passwordField).toHaveValue('abcdefghjklmnopqrstuvwxyz')
		);

		fireEvent.blur(passwordField);

		registerSetup.getByText(
			'Your password must be a maximum of 20 characters'
		);
	});

	it('should error if submitted without valid password', async () => {
		const { registerSetup, emailField, passwordField, submitButton } =
			setup();

		expect(emailField).toBeInTheDocument();
		expect(passwordField).toBeInTheDocument();

		await act(async () => {
			fireEvent.click(submitButton);
		});

		registerSetup.getByText('You must enter a valid password');
	});

	// it('should error on empty name on blur', async () => {
	// 	const {
	// 		registerSetup,
	// 		firstNameField,
	// 		lastNameField,
	// 		emailField,
	// 		passwordField
	// 	} = setup();

	// 	expect(firstNameField).toBeInTheDocument();
	// 	expect(lastNameField).toBeInTheDocument();
	// 	expect(emailField).toBeInTheDocument();
	// 	expect(passwordField).toBeInTheDocument();

	// 	fireEvent.focus(firstNameField);
	// 	userEvent.type(firstNameField, ' ');
	// 	await waitFor(() => expect(firstNameField).toHaveValue(' '));

	// 	fireEvent.blur(firstNameField);

	// 	registerSetup.getByText('You must enter a first name');
	// });

	// it('should error on empty name on blur', async () => {
	// 	const {
	// 		registerSetup,
	// 		firstNameField,
	// 		lastNameField,
	// 		emailField,
	// 		passwordField
	// 	} = setup();

	// 	expect(firstNameField).toBeInTheDocument();
	// 	expect(lastNameField).toBeInTheDocument();
	// 	expect(emailField).toBeInTheDocument();
	// 	expect(passwordField).toBeInTheDocument();

	// 	fireEvent.focus(firstNameField);
	// 	userEvent.type(firstNameField, ' ');
	// 	await waitFor(() => expect(firstNameField).toHaveValue(' '));

	// 	fireEvent.blur(firstNameField);

	// 	registerSetup.getByText('You must enter a first name');
	// });

	it('should error on empty email on blur', async () => {
		const {
			registerSetup,
			firstNameField,
			lastNameField,
			emailField,
			passwordField
		} = setup();

		expect(firstNameField).toBeInTheDocument();
		expect(lastNameField).toBeInTheDocument();
		expect(emailField).toBeInTheDocument();
		expect(passwordField).toBeInTheDocument();

		fireEvent.focus(emailField);
		userEvent.type(emailField, 'j@gmail');
		await waitFor(() => expect(emailField).toHaveValue('j@gmail'));

		fireEvent.blur(emailField);

		registerSetup.getByText('You must enter a valid email address');
	});

	it('should error on empty password on blur', async () => {
		const {
			registerSetup,
			firstNameField,
			lastNameField,
			emailField,
			passwordField
		} = setup();

		expect(firstNameField).toBeInTheDocument();
		expect(lastNameField).toBeInTheDocument();
		expect(emailField).toBeInTheDocument();
		expect(passwordField).toBeInTheDocument();

		fireEvent.focus(passwordField);
		userEvent.type(passwordField, 'pwd');
		await waitFor(() => expect(passwordField).toHaveValue('pwd'));

		fireEvent.blur(passwordField);

		registerSetup.getByText(
			'Your password must be a minimum of 6 characters'
		);
	});

	// it('should make social sign up request', async () => {
	// 	const { socialSignup } = setup();

	// 	await act(async () => {
	// 		fireEvent.click(socialSignup);
	// 	});

	// 	expect(mockSocialRegister).toHaveBeenCalledTimes(1);
	// });
});
