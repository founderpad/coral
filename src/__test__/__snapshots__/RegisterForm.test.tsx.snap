// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register form matches snapshot 1`] = `
<DocumentFragment>
  <form
    aria-label="registerForm"
    autocomplete="off"
    id="registerForm"
    name="registerForm"
    novalidate=""
  >
    <div
      class="chakra-stack css-w4pe59"
    >
      <div
        class="chakra-form-control css-1naxzcr"
        role="group"
      >
        <input
          aria-label="firstName"
          aria-required="true"
          class="chakra-input css-wv9m3t"
          control="[object Object]"
          id="firstName"
          name="firstName"
          placeholder="First name"
          required=""
          rules="[object Object]"
          value=""
        />
      </div>
      <div
        class="chakra-form-control css-1naxzcr"
        role="group"
      >
        <input
          aria-label="lastName"
          class="chakra-input css-wv9m3t"
          control="[object Object]"
          id="lastName"
          name="lastName"
          placeholder="Last name"
          rules="[object Object]"
          value=""
        />
      </div>
      <div
        class="chakra-form-control css-1naxzcr"
        role="group"
      >
        <input
          aria-label="email"
          aria-required="true"
          class="chakra-input css-wv9m3t"
          control="[object Object]"
          id="email-email"
          name="email"
          placeholder="Email"
          required=""
          rules="[object Object]"
          type="email"
          value=""
        />
      </div>
      <div
        class="chakra-form-control css-1naxzcr"
        role="group"
      >
        <input
          aria-label="password"
          aria-required="true"
          autocomplete="current-password"
          class="chakra-input css-wv9m3t"
          control="[object Object]"
          id="password-password"
          name="password"
          placeholder="Password"
          required=""
          rules="[object Object]"
          type="password"
          value=""
        />
      </div>
      <button
        aria-label="submit-register-account"
        class="chakra-button css-1pqrlwa"
        disabled=""
        id="submit-register-account"
        type="submit"
      >
        Create account
      </button>
    </div>
  </form>
  <div
    class="chakra-text css-i2jbli"
  >
    By continuing, you agree to founderpad's 
    <span
      class="chakra-link css-ntv2za"
      href="/terms-of-conditions"
      rel="noopener noreferrer"
      target="_blank"
      title="Link to Terms of Service"
    >
      Terms of Service
    </span>
     and acknowledge that you've read our 
    <span
      class="chakra-link css-ntv2za"
      href="/privacy-policy"
      rel="noopener noreferrer"
      target="_blank"
      title="Link to Privacy Policy"
    >
      Privacy Policy.
    </span>
     
  </div>
  <div
    class="chakra-text css-1gx0chr"
  >
    Already have an account? 
    <span
      class="chakra-link css-ntv2za"
      href="/login"
      title="Link to login"
    >
      Login here
    </span>
  </div>
</DocumentFragment>
`;
